### PR TITLE
Homework 3

### PR DESCRIPTION
**Pull request by kieracawley**

### CODE REVIEW

File: cloudObjectManager.js
Comment (rickyc) [inline cloudObjectManager.js]: Blank constructors are not necessary.

File: cloudObjectManager.js
Comment (rickyc) [inline cloudObjectManager.js]: We talked about this already, but to document it. The store class should probably have a variable called `this.products = []`.

File: cloudObjectManager.js
Comment (rickyc) [inline cloudObjectManager.js]: ES6.

`(list) => {`

File: cloudObjectManager.js
Comment (rickyc) [inline cloudObjectManager.js]: Your indentation is all over the place.
http://stackoverflow.com/questions/22608160/atom-change-indentation-mode

File: objectVisualManager.js
Comment (rickyc) [inline objectVisualManager.js]: Classes start with a capital.

Comment (rickyc) [reply objectVisualManager.js]: ObjectVisualManager

File: objectVisualManager.js
Comment (rickyc) [inline objectVisualManager.js]:
``` js
let image = `<img class='productImage' src='${product._image_url}>`;
```

File: script.js
Comment (rickyc) [inline script.js]: Here you can call `newStore.fetchProducts()`. You can pass it a callback function so when it's done fetching products, it can update the UI.

File: store.js
Comment (rickyc) [inline store.js]: `const`

File: store.js
Comment (rickyc) [inline store.js]: `const`

File: store.js
Comment (rickyc) [inline store.js]: `const`. All these are `const` since you never assign (use equal sign) it to something else.

File: store.js
Comment (rickyc) [inline store.js]: es6 strings

File: store.js
Comment (rickyc) [inline store.js]: `isOpen`. Wrong language :). CamelCase with JavaScript.

**Review by rickyc — CHANGES_REQUESTED** [review]
Your indentation is a bit all over the place.

**Comment by rickyc** [comment]
Did you check your work? 🤔 . I get this error on load 😭 .

error on load

**Comment by rickyc** [comment]
@thebluecat5445 Comments were never addresses, you can not merge this one until you fix it.

**Comment by kieracawley** [comment]
@rickyc I'm still working on the sign in and the comments so it might be broken

**Comment by rickyc** [comment]
Ahh, got it! I forgot to add a `STATUS` on the PR template. We usually write `WIP` (work in progress) so people know it's not complete. We also use the labels feature to help with that.

File: Homework 3/ObjectVisualManager.js
Comment (rickyc) [inline Homework 3/ObjectVisualManager.js:2]: Blank constructor isn't necessary. Houndci would also complain about the formatting :)

File: Homework 3/ObjectVisualManager.js
Comment (rickyc) [inline Homework 3/ObjectVisualManager.js]: This is actually something built into JS (and most languages), but good job writing it yourself :). It's good to know how to do it.

http://www.w3schools.com/jsref/jsref_tofixed.asp

File: Homework 3/script.js
Comment (rickyc) [inline Homework 3/script.js]: Instead of leaving a blank if statement you can just do `if (!user.isStoreOwner)` instead of the `else`. `_isStoreOwner` is private so you should never call that from a class object. You need to call the public method which has no underscore.

File: Homework 3/script.js
Comment (rickyc) [inline Homework 3/script.js:151]: You can actually chain multiple CSS selectors together.

``` js
$("#mainStoreOwnerPage, #signInPage").hide();
```

File: Homework 1/script.js
Comment (rickyc) [inline Homework 1/script.js:3]: You should take a look at how I refactored it, but you are close!

https://nanohackeracademy.slack.com/files/ricky/F2R5P1BM5/refactored_js.js

Basically you want an array of strings in this case and just loop through those.

File: Homework 2/script.js
Comment (rickyc) [inline Homework 2/script.js:52]: Let's use [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) here.

``` js
const a = [1,2,3,4,5,6].map((i) => {
  return i * 2;
});

console.log(a);

// this returns [2,4,6,8,10,12]
```

File: Homework 3/Store.js
Comment (rickyc) [inline Homework 3/Store.js:3]: Fix your indentation, you don't want Joe code.

File: Homework 3/Store.js
Comment (rickyc) [inline Homework 3/Store.js:27]: ES6 strings...

File: Homework 3/User.js
Comment (rickyc) [inline Homework 3/User.js:20]: ES6 function syntax

**Review by rickyc — CHANGES_REQUESTED** [review]

**Comment by kieracawley** [comment]
@rickyc I think I got everything

File: Homework 3/ObjectVisualManager.js
Comment (rickyc) [inline Homework 3/ObjectVisualManager.js:5]: 👍

File: Homework 3/ObjectVisualManager.js
Comment (rickyc) [inline Homework 3/ObjectVisualManager.js:8]: Using ` actually lets you do multiline!

```js
const str = `
multiline
yay
`
```

File: Homework 3/ObjectVisualManager.js
Comment (rickyc) [inline Homework 3/ObjectVisualManager.js:13]: Hmm, you are referencing a private variable. `_` means private so it should never be accessed with a `.`. It's confusing since it works in JS, but it actually doesn't work in any other language. If you need to access costInCents, it should be with a public method.

`product.costInCents`

File: Homework 3/ObjectVisualManager.js
Comment (rickyc) [inline Homework 3/ObjectVisualManager.js:25]: This is scoped to the function on line 23, so you can actually call it `manager`.

File: Homework 3/ObjectVisualManager.js
Comment (rickyc) [inline Homework 3/ObjectVisualManager.js:26]: A get function usually means it return something. In this case it looks like it's actually doing more than that. Consider renaming the function to something that describes what it's doing.

File: Homework 3/Store.js
Comment (rickyc) [inline Homework 3/Store.js:17]: You set `numberofAddedProducts` to 0. 🤔 Doesn't that mean you just want to do `if (numberofProducts === 0)`?

File: Homework 3/Store.js
Comment (rickyc) [inline Homework 3/Store.js:20]: You should use a for loop instead. That gets rid of the x on line 14 and the line 22.

File: Homework 3/index.html
Comment (rickyc) [inline Homework 3/index.html:69]: 😱 this is Joe code. @blim8183 look at that `href` in a `<p>`!!!

File: Homework 3/index.html
Comment (rickyc) [inline Homework 3/index.html:73]: 🚫 href should only be on <a> tags. Use an <a> tag instead and just style it as a button using bootstrap.

File: Homework 3/index.html
Comment (rickyc) [inline Homework 3/index.html:87]: Wrap inputs in a `<form>` tag. I skipped it in the earlier classes mainly since button clicks will perform a POST on the form therefore refreshing the page. You'll need to do `event.preventDefaults()` to prevent it from refreshing (aka default behavior).

File: Homework 3/index.html
Comment (rickyc) [inline Homework 3/index.html:89]: What is the `eventcharCode` numbers? Inline code is no good since it's hard to trace.

File: Homework 3/script.js
Comment (rickyc) [inline Homework 3/script.js:22]:
```js
const bool = ["#productNameInput", "#productImageInput", "#productCostInput"].some((id) => $(id).val() === "");
```

I think this will work.

File: Homework 3/script.js
Comment (rickyc) [inline Homework 3/script.js:27]: Again there's some code duplication. If you abstract the above in line 22 into it's own array, you can reuse it. Actually, you can probably just refactor and use a class. Don't think `.value` works actually. That's vanilla js.

```js
$("#productNameInput, #productCostInput, #productImageInput").val("")
```

You can chain css selectors together.

File: Homework 3/script.js
Comment (rickyc) [inline Homework 3/script.js:56]: if statements should really never be this deep. That typically means you can probably refactor it. If you were to step back and read this code in a few months, you'd probably be confused.

File: Homework 3/script.js
Comment (rickyc) [inline Homework 3/script.js:86]: Same error message and function is showing up over and over. Copy and paste might make sense now, but what if I told you to refactor the `else` behavior? Now you have to copy and paste everywhere. If the logic gets more complicated, you'll end up spending more time looking for the places to update.

File: Homework 3/script.js
Comment (rickyc) [inline Homework 3/script.js:100]: Again css selectors can be chained. Study up CSS selectors a bit. http://www.w3schools.com/cssref/css_selectors.asp

```js
$("#signOutButton, #ownerSignOutButton")
```

File: Homework 3/style.css
Comment (rickyc) [inline Homework 3/style.css:29]:
```css
#pageTitle, #ownerPageTitle {
  font-size: 25px;
}
```

However it's probably better to use a class `page-title`.

File: Homework 3/style.css
Comment (rickyc) [inline Homework 3/style.css:170]: You're duplicating css right now. You can just make a class for similar styles and then make an ID to overwrite specifics.

File: Homework 3/script.js
Comment (rickyc) [inline Homework 3/script.js:140]: You can use `const` in a for loop instead of a `let`.

File: Homework 3/CloudObjectManager.js
Comment (rickyc) [inline Homework 3/CloudObjectManager.js:147]: Use ES6 function syntax.

```js
newRelation.save({
  success: (obj) => {
    callBack();
  },
  error: (err) => {
  }
});
```

This can actually be refactored more. I think you can actually call callback directly since it's function.

```js
newRelation.save({
  success: callback,
  error: (err) => { }
});
```

I believe you can actually do the following in ES6. It's shorthand to represent the same thing as above.

```js
newRelation.save({
  success(obj) {
    callBack();
  },
  error(err) {
  }
});
```

**Review by rickyc — CHANGES_REQUESTED** [review]
Left some comments on the code. I didn't completely mark every line where there were problems since a lot of it repeats. You should take a pass through and make changes yourself.

I haven't looked at the app itself yet. Will do that tomorrow.

**Comment by rickyc** [comment]
From a UI/UX perspective. It wasn't too bad. I was able to figure out how to use the site with relative ease. One thing to watch for are error states. I put in the same username and the same email and both were apparently violating the unique constraint on sign up. No error was thrown so I was confused. Good error messages on login. I put in my email first and it threw an error that was clear.

After you create a new product, you should typically display a success message and clear the fields. I think not having the success message is okay since I was able to see the product automatically updated into the "Your Products." 

Refreshing the page made me lose where I was :(. You should be using the CloudBoost User to persist login state. Typically, a user should not be prompted to re-login unless they sign out.

For the Store part, notice how you have to scroll within a box once you add 3 items. What if there were 100 items? It would be impossible to find things. Looks like clicking on the huge box doesn't actually add or remove an item, so it might be better to just use a list format. 

The buttons are a bit destructive. That means if I hit the remove by accident it actually removes it without prompting me to see if I actually want to delete it. Delete actions should always have prompts as the user doesn't know what to expect until they do it for the first time.

Overall it works and it didn't take me too long to use, so 👍

File: Homework 3/index.html
Comment (blim8183) [reply Homework 3/index.html:69]: G8, good job Joe.

**Pull request merged**
